runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -214,21 +213,22 @@ State waitUntilFinish(
     MonitoringUtil monitor = new MonitoringUtil(projectId, dataflowOptions.getDataflowClient());
 
     long lastTimestamp = 0;
-    BackOff backoff =
-        duration.getMillis() > 0
-            ? new AttemptAndTimeBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS,
-                MESSAGES_POLLING_INTERVAL,
-                duration.getMillis(),
-                AttemptAndTimeBoundedExponentialBackOff.ResetPolicy.ATTEMPTS,
-                nanoClock)
-            : new AttemptBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS, MESSAGES_POLLING_INTERVAL);
+    FlexibleBackoff backoff = FlexibleBackoff.of()
+        .withMaxRetries(MESSAGES_POLLING_RETRIES)
+        .withInitialBackoff(MESSAGES_POLLING_INTERVAL);
+    if (duration.isLongerThan(Duration.ZERO)) {
+      backoff = backoff.withMaxCumulativeBackoff(duration);
+    }
+
+    // In this function, track the cumulative time since we've started backing off and use this
+    // to enforce a wall-clock limit.
+    long startNanos = nanoClock.nanoTime();
```

Comment:
Ah, looking at more of the context, yeah. This comment should be:

```
// This function tracks the cumulative time from the *first request* to enforce the wall-clock limit.
// Any backoff instance could, at best, track the the time since the first attempt at a given request.
// Thus, we need to track the cumulative time ourselves.
```